File: prueba-tecnica-backend/src/main/java/com/lautaro_cenizo/service/HistorialServicioService.java
package com.lautaro_cenizo.service;

import com.lautaro_cenizo.crud.dto.HistorialServicioDTO;
import com.lautaro_cenizo.entity.historial.HistorialServicio;

import java.util.List;
import java.util.Optional;

public interface HistorialServicioService {
    HistorialServicio crear(HistorialServicioDTO historialDTO);
    HistorialServicio actualizar(Integer id, HistorialServicioDTO historialDTO);
    Optional<HistorialServicio> obtenerPorId(Integer id);
    List<HistorialServicio> listarTodos();
    List<HistorialServicio> listarPorCliente(Integer clienteId);
    List<HistorialServicio> listarPorTecnico(Integer tecnicoId);
    void eliminar(Integer id);
    void calificarServicio(Integer id, int calificacion);
    void marcarControlCalidad(Integer id, boolean aprobado);
}

File: prueba-tecnica-frontend/src/app/service/historial-servicio.service.ts
import { Injectable } from '@angular/core';
import { HistorialServicio } from '../models/class/historialServicio';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { HistorialServicioDTO } from '../models/dto/historialServicioDto';

@Injectable({
  providedIn: 'root'
})
export class HistorialServicioService {

  private baseUrl = 'http://localhost:8080/api/historial-servicios';

  constructor(private http: HttpClient) { }

  crear(historialDTO: HistorialServicioDTO): Observable<HistorialServicio> {
    return this.http.post<HistorialServicio>(this.baseUrl, historialDTO);
  }

  actualizar(id: number, historialDTO: HistorialServicioDTO): Observable<HistorialServicio> {
    return this.http.put<HistorialServicio>(`${this.baseUrl}/${id}`, historialDTO);
  }

  obtenerPorId(id: number): Observable<HistorialServicio> {
    return this.http.get<HistorialServicio>(`${this.baseUrl}/${id}`);
  }

  listarTodos(): Observable<HistorialServicio[]> {
    return this.http.get<HistorialServicio[]>(this.baseUrl);
  }

  listarPorCliente(clienteId: number): Observable<HistorialServicio[]> {
    return this.http.get<HistorialServicio[]>(`${this.baseUrl}/cliente/${clienteId}`);
  }

  listarPorTecnico(tecnicoId: number): Observable<HistorialServicio[]> {
    return this.http.get<HistorialServicio[]>(`${this.baseUrl}/tecnico/${tecnicoId}`);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  calificarServicio(id: number, calificacion: number): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${id}/calificar`, null, {
      params: { calificacion: calificacion.toString() }
    });
  }

  marcarControlCalidad(id: number, aprobado: boolean): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${id}/control-calidad`, null, {
      params: { aprobado: aprobado.toString() }
    });
  }

}
